### DISCUSSION TITLE
Lotus钱包常用操作及钱包使用场景？

### DISCUSSION THREAD

**Discussion by maxvint** [filecoin-project/community-china#15] [新手上路]
# Lotus钱包管理及使用场景

## 1. Lotus钱包常用操作
列出钱包地址:
```sh
$ lotus wallet list
```

查看余额:
```sh
$ lotus wallet balance <WALLET_ADDRESS>
```

默认钱包地址:
```sh
$ lotus wallet default
```

设置默认钱包地址:
```sh
$ lotus wallet set-default <WALLET_ADDRESS>
# 例如: lotus wallet set-default fxxxx001
```

从默认钱包地址发送代币:
```sh
$ lotus send <TARGET_ADDRESS> <AMOUNT>
# 例如: lotus send fxxxxx001 10
```

从指定钱包地址发送代币:
```sh
$ lotus send --from=<SENDER_ADDRESS> <TARGET_ADDRESS> <AMOUNT>
# 例如: lotus send --from=fxxxxxx002 fxxxxxx001 3
```

导出钱包和私钥:
```sh
$ lotus wallet export <WALLET_ADDRESS>
# 例如: lotus wallet export fxxxxx001 > fxxxxx001.privatekey
```

通过私钥导入钱包:
```sh
$ lotus wallet import 
# 根据提示输入私钥，导入钱包
# 例如: lotus wallet import fxxxxx001.privatekey
```

## 2. 为Miner配置多个钱包
一个Miner可以配置：Owner地址、Worker地址、Control地址。

![wallet](https://user-images.githubusercontent.com/1715211/110258632-8d98ca00-7fde-11eb-9c09-2d7308b4fad2.png)

Miner初始化的时候，会创建一个`f0`开头的Actor地址并发送到区块链上，所有发送到Miner的转账消息，都会发送到Actor账户中。
可以为Miner配置以下几个不同类型的钱包，分别负责提现、时空证明、PreCommitSector、ProveCommitSector。

#### 2.1 Owner钱包
Miner初始化的时候，会绑定一个Owner钱包，Owner钱包主要用于：
- 更换Miner的Owner或Worker钱包；
- Miner Actor提现后的币将进入Owner钱包；
- 当WindowPoSt钱包账户余额不足时，用于提交时空证明；

可通过以下命令更换Owner地址（感谢@xiaoyuaiheshui更新的最新命令）：
```sh
$ lotus-miner actor set-owner --really-do-it [new f3address] [old f3address]
$ lotus-miner actor set-owner --really-do-it [new f3address] [new f3address]
```
更换后的新Owner钱包地址，也需要导入到Miner连接的Daemon节点中。

#### 2.2 Worker钱包
Worker钱包用于发送和支付矿工执行的日常操作，包括：
- 初始化Miner；
- 更换Miner的`peer id`或`multiaddresses`；
- 与订单市场或支付通道的actor交互；
- 签名新的区块；
- 提交证明消息、上报失败扇区消息；
- 在下列情况下，用来提交时空证明的消息：
  - 没有配置时空证明的Control地址，或时空证明的Control地址余额不足；
  - Owner钱包地址余额不足。

Worker钱包必须保持有充足的币，否则将会影响封装、时空证明等。

#### 2.3 Control钱包
Control钱包用来提交时空证明，以及提交PreCommit和ProveCommit消息。

**(1) 创建Control钱包**
```sh
lotus wallet new bls
> f3defg...

lotus send --from <address> f3defg... 100
```
**(2) 配置Control钱包**
```sh
lotus-miner actor control set --really-do-it f3defg...

> Add f3defg...
> Message CID: bafy2...
```

**(3) 等待消息上链**
```sh
lotus state wait-msg bafy2...

> ...
> Exit Code: 0
> ...
```

**(4) 查看Control列表**
```sh
lotus-miner actor control list

> name       ID      key        use    balance
> owner      t01111  f3abcd...  other  300 FIL
> worker     t01111  f3abcd...  other  300 FIL
> control-0  t02222  f3defg...  post   100 FIL
```

## 3. 为PreCommit和ProveCommit配置Control钱包
**(1) 创建Control钱包**

通过以下命令，创建2个f3类型的钱包地址，并给钱包转币。
```sh
lotus wallet new bls

> f3rht...

lotus wallet new bls

> f3sxs...

lotus wallet list

> Address   Balance  Nonce  Default
> f3rht...  0 FIL    0      X
> f3sxs...  0 FIL    0
```
**(2) 获取钱包ID**
```sh
lotus wallet list -i

 > Address   ID        Balance                   Nonce  Default
 > f3rht...  f0100933  0.59466768102284489 FIL   1      X
 > f3sxs...  f0100939  0.4 FIL                   0
```
**(3) 添加Control地址**
```sh
lotus-miner actor control set --really-do-it=true f0100933 f0100939

 > Add f3rht...
 > Add f3sxs...
 > Message CID: bafy2bzacecfryzmwe5ghsazmfzporuybm32yw5q6q75neyopifps3c3gll6aq

lotus actor control list

 > name       ID      key        use    balance
 > owner      t01...  f3abcd...  other  15 FIL
 > worker     t01...  f3abcd...  other  10 FIL
 > control-0  t02...  f3defg...  post   100 FIL
 > control-1  t02...  f3defg...  post   100 FIL
```
**(4) 配置Control地址**
修改Miner的配置文件`$LOTUS_MINER_PATH/config.toml`，更新`[Addresses]`中的`PreCommitControl`和`CommitControl`地址。
```toml
[Addresses]
PreCommitControl = ["f3rht..."]
CommitControl = ["f3sxs..."]
```

**Comment by rocklau** [comment]
$ lotus-miner actor set-owner --really-do-it <WALLET_ADDRESS>

现在无效了。 有新的方法吗？我怎么设置都不行。

$ lotus-miner actor set-owner --really-do-it [newf3address] [oldf3address]

这个也不行

**Reply by xiaoyuaiheshui** [reply]
更改owner要执行两次
 lotus-miner actor set-owner --really-do-it [newf3address] [oldf3address]
 lotus-miner actor set-owner --really-do-it [newf3address] [newf3address]

**Comment by dcshy** [comment]
学习了
自有技术团队近百人， 机房密封日产能2p+ ， 独立节点集群方案， 可加vx：IPFS-Shy

**Comment by galaxy0x** [comment]
Owner钱包强烈建议更换为冷钱包或硬件钱包，可通过以下命令更换Owner地址。
更换后的新Owner钱包地址，也需要导入到Miner连接的Daemon节点中。
如果owner改成冷钱包地址 又导入到daemon节点 那不马上变成热钱包了 这一段怎么理解

**Reply by 0xhhhhh** [reply]
同问

**Reply by galaxy0x** [reply]
搞清楚了 更换其他钱包要用老owner私钥签名 执行一次
lotus-miner actor set-owner --really-do-it [newf3address] [oldf3address]
再用新私钥签名一次
lotus-miner actor set-owner --really-do-it [newf3address] [newf3address]
更换完成后新owner私钥是可以不用放在节点的。

**Comment by KaixuanZhao** [comment]
RRMine，雅典娜矿池，全网算力领先1000PB，需要B端C端方案的滴滴我，交流学习，欢迎加入社群Wechet：kkmmjjnn00

**Comment by dcshy** [comment]
filecoin存量迁移出海 新增量海外机房上架封装 欢迎合作交流 v：：IPFS-Shy

**Comment by zhaobo443** [comment]
FIL联合封装，用户提供质押币，公司提供设备和运维，收益60％,公司扣出20％服务费和20%设备费，合同期1080天，540天质押币退还，t3天封装。可实地考察。V:zhaobo9308
